### ISSUE TITLE
Clearing promnesia's database safely?

### ISSUE THREAD

**Issue by Cobertos** [karlicoss/promnesia#210]
I tried adding my entire `/projects` directory using an `auto` indexer but found I created a 33GB database by accident. I deleted it and started over but was wondering on the best testing workflow going forward

* What is your workflow for testing stuff that might change how the database is built, if you don't just delete the database and regenerate it already?
* What am I losing by deleting the database? The database is just an index (right?) and the source of truth is elsewhere (like for .md files). Does this hold true for things like browser visits as well? If it does not, is there and exporter for something like HPI to use a Firefox export as the source of truth? Am I losing anything else?
* Is there a 'dry-run' option or something for `promnesia index`, so I can see what it will add and perhaps the sizes?

**Comment by karlicoss** [comment]
> I tried adding my entire /projects directory using an auto indexer but found I created a 33GB database by accident.

Whoops... but good catch. Maybe we should throw an exception if the database gets too big (say >10Gb), and allow the user to tweak the threshold in config if they are sure. Would be sad to fill up someone's disk.

> What am I losing by deleting the database? The database is just an index (right?) 

Yep! In fact, by default it's overwriting everything anyway
https://github.com/karlicoss/promnesia/blob/21181a862820bd0ffa59f888cd780602221591ea/src/promnesia/dump.py#L17-L22

So it's just an index, and most likely will be this way. In the future it might make sense to support different backends for performance (e.g. postgre), and migrating people's sources of true data risking losing it is definitely not something I want to be responsible for :)

> If it does not, is there and exporter for something like HPI to use a Firefox export as the source of truth? Am I losing anything else?

Yeah, I still need to expose the browser history to HPI, perhaps (it kind of predated it so that's why ended up in Promnesia).
I'm using [this script](https://github.com/karlicoss/promnesia/blob/20a07c23d150e2206f22d1afa5b1a4f1a17e3305/scripts/browser_history.py#L107-L117) for regular snapshots of Firefox history, which then are used by promnesia. Also just had relevant discussion here: https://github.com/karlicoss/promnesia/pull/207

> Is there a 'dry-run' option or something for promnesia index, so I can see what it will add and perhaps the sizes?

I guess the closest is something like `promnesia demo --no-serve /path/to/files` -- it would put the database in a tmp file and allow you to inspect (cleaning up when you press enter). But that wouldn't save from overfilling the disk.

I guess an actual dry run mode where it just goes through everything and estimates database size would make some sense, yeah. Although maybe the threshold thing above would be enough.


P.S. by the way if you're heavily indexing plaintext files, you might benefit from `PROMNESIA_CORES`
https://github.com/karlicoss/promnesia/blob/21181a862820bd0ffa59f888cd780602221591ea/src/promnesia/config.py#L128-L134
It's been hidden behind the env variable since I was just experimenting, but soon I'll probably expose it properly to the cli or something.

**Comment by Cobertos** [comment]
Not so much worried about disk fill as I am just being able to see how much is going in and cleaning it up easily after. For end users an exception or warning might be nice for disk fillage, if they're getting close to their filesystem usage maybe. Definitely edge-case

Might not even need a dry run mode, but just the filesize added to the index in the debug log would be enough for me. I could at least then Ctrl + C, add the filter if a particularly large JSON file gets in, delete the index, start over.

In the future, it would be really nice for the VCS indexer to listen to .gitignore or similar files. I noticed when it was scanning some Unity projects, there's a lot of cache-type folders it was finding a lot of big weird json files in with nothing useful, but also no easy way to prune them without just manually ignoring the folder.

But okay, I will experiment with the PROMNESIA_CORES, `promnesia demo --no-serve /path/to/files`, and the firefox HPI script thing. **And it sounds like deleting my index is no issue, so that solves this.**

One thing though, does 

> Yeah, I still need to expose the browser history to HPI, perhaps (it kind of predated it so that's why ended up in Promnesia).

mean that eventually all the of the data gathering will come from HPI? I was wondering about that. It seems like if HPI was the big collector and exposer but Promnesia was more a single interface for exploration into data, though Promnesia currently does some data gathering?

Also is there a Discord or IRC or _something_ for discussion about this stuff? Might be easier than a GitHub issue x3

**Comment by karlicoss** [comment]
> but just the filesize added to the index in the debug log would be enough for me.

Yeah, makes sense! Guess it would belong somewhere here: https://github.com/karlicoss/promnesia/blob/e44a2bda053cc93bb8dc1b84fd1a4461aa5d914f/src/promnesia/dump.py#L74, could log every 10K visits.. or something like that. For database size probably best to use some sqlite native function (instead of st_size or something) to make sure the journal is accounted for.

>  the future, it would be really nice for the VCS indexer to listen to .gitignore or similar files.

So, fdfind [already does it](https://github.com/sharkdp/fd#excluding-specific-files-or-directories)! And it also supports `.ignore` files, which you can use to ignore stuff that can't be put in `.gitignore`. Wouldn't make much sense to reimplement it for `find` (considering fdfind is available in most distros repositories), but I guess could add a suggestion to use it to `promnesia doctor` or something. 

But again, you somehow keep discovering all experimental features in one go haha. There is also `PROMNESIA_FD_EXTRA_ARGS` env variable which is passed to `fdfind` and used as an extra exclude list (so you don't have to scatter them across the filesystem). E.g. I have something like `PROMNESIA_FD_EXTRA_ARGS=--ignore-file /promnesia-extra/ignorefile`.

> mean that eventually all the of the data gathering will come from HPI

Yeah, probably? HPI is more of an 'interface', I'm trying to keep the exporters/collectors/parsers ([and even individual 'interfaces' that interpret data](https://beepb00p.xyz/exports.html#dal)) separate as far as possible to make sure it's all doesn't end up as a monolith. I'd prefer that other people can benefit from data exporters/parsers without necessarily having to use HPI, at least to the maximum extent that feasible.

> though Promnesia currently does some data gathering?

In case of browser, it's more of 'scripts happen to be in Promnesia repository', but they aren't called from Promnesia, so in that sense they are decoupled. 

Basically, the 'problematic' bit here is that the [code that extracts URLs](https://github.com/karlicoss/promnesia/blob/master/src/promnesia/sources/browser.py#L136-L284) is 'siloed' inside `promnesia.browser` module. So conceptually would be nice to expose them as an HPI module. The main reason I haven't yet is that browser history hasn't been really useful for me anywhere else apart from Promnesia, so it hasn't been of high priority. 

> Also is there a Discord or IRC or something for discussion about this stuff? Might be easier than a GitHub issue x3

Yeah indeed.. I've been thinking about setting something up for a while. There are a couple of channels dedicated to it on others' discords, e.g. you're welcome to [this one](https://discord.gg/eryJRbgE)!

**Comment by Cobertos** [comment]
> So, fdfind already does it!

Oh nice! Holy shit, that beats the garbage I was trying to do with [find](https://github.com/Cobertos/dotfiles/blob/master/cobertos.bashrc#L51). I will have to fully switch over.

> but I guess could add a suggestion to use it to promnesia doctor or something.

I guess I had no idea until I saw it called out in the code (switching from Windows to Linux) and then looked it up and realized how many stars/friend's stars it had. Maybe it's something more knowledgable Linux people know about but I'm just new to this OS uwu

> browser history hasn't been really useful for me anywhere else apart from Promnesia

Makes sense, I think I assumed that bookmark retrieval and history retrieval (due to permissions request) were happening in the plugin and not something generally exposed by HPI necessarily. It's not of immediate useful to me either.

> e.g. you're welcome to this one!

I think the link ded :c
